contracts/StakeManagerV2.sol

```diff
@@ -0,0 +1,343 @@
+        if (_staderTreasury != address(0)) revert ZeroAddress();
+        staderTreasury = _staderTreasury;
+        emit SetStaderTreasury(staderTreasury);
+    }
+
+    function setFeeBps(uint256 _feeBps) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        feeBps = _feeBps;
+        emit SetFeeBps(feeBps);
+    }
+
+    /// @notice Delegate BNB to the preferred operator.
+    /// @param _referralId referral id of KOL
+    /// @return The amount of BnbX minted.
+    function delegate(string calldata _referralId)
+        external
+        payable
+        override
+        whenNotPaused
+        nonReentrant
+        returns (uint256)
+    {
+        uint256 amountToMint = convertBnbToBnbX(msg.value);
+        _delegate();
+        BNBX.mint(msg.sender, amountToMint);
+
+        emit DelegateReferral(msg.sender, msg.value, amountToMint, _referralId);
+        return amountToMint;
+    }
+
+    /// @notice Request to withdraw BnbX and get BNB back.
+    /// @param _amount The amount of BnbX to withdraw.
+    /// @return The index of the withdrawal request.
+    function requestWithdraw(uint256 _amount) external override whenNotPaused nonReentrant returns (uint256) {
+        if (_amount == 0) revert ZeroAmount();
+
+        withdrawalRequests.push(
+            WithdrawalRequest({ user: msg.sender, amountInBnbX: _amount, claimed: false, batchId: 0, processed: false })
+        );
+
+        uint256 requestId = withdrawalRequests.length - 1;
+        userRequests[msg.sender].push(requestId);
+
+        BNBX.transferFrom(msg.sender, address(this), _amount);
+        emit RequestedWithdrawal(msg.sender, _amount);
+
+        return requestId;
+    }
+
+    /// @notice Start the batch undelegation process.
+    /// @param _batchSize The size of the batch.
+    /// @param _operator The address of the operator to undelegate from.
+    function startBatchUndelegation(
+        uint256 _batchSize,
+        address _operator
+    )
+        external
+        override
+        whenNotPaused
+        nonReentrant
+        onlyRole(OPERATOR_ROLE)
+    {
+        if (_operator == address(0)) {
+            // if operator is not provided, use preferred operator
+            _operator = OPERATOR_REGISTRY.preferredWithdrawalOperator();
+        }
+        if (!OPERATOR_REGISTRY.operatorExists(_operator)) {
+            revert OperatorNotExisted();
+        }
+
+        address creditContract = STAKE_HUB.getValidatorCreditContract(_operator);
+        uint256 pooledBnb = IStakeCredit(creditContract).getPooledBNB(address(this));
+
+        uint256 cummulativeBnbXToBurn;
+        uint256 processedCount;
+        uint256 amountInBnbXToBurn;
+        while (firstUnprocessedUserIndex < withdrawalRequests.length && processedCount < _batchSize) {
+            cummulativeBnbXToBurn = amountInBnbXToBurn + withdrawalRequests[firstUnprocessedUserIndex].amountInBnbX;
+            if (pooledBnb >= convertBnbXToBnb(cummulativeBnbXToBurn)) {
```

Comment:
This can go into the while condition as well. Will leave the readability trade-off to you. 

The current code is not easy on the eye.

Reply:
refactored in #78  . please review it carefully